instrumentation/log4j/log4j-2.7/src/main/java/io/opentelemetry/instrumentation/auto/log4j/v2_7/Log4j27Instrumentation.java

```diff
@@ -0,0 +1,86 @@
+ * limitations under the License.
+ */
+
+package io.opentelemetry.instrumentation.auto.log4j.v2_7;
+
+import static io.opentelemetry.javaagent.tooling.ClassLoaderMatcher.hasClassesNamed;
+import static io.opentelemetry.javaagent.tooling.bytebuddy.matcher.AgentElementMatchers.implementsInterface;
+import static net.bytebuddy.matcher.ElementMatchers.isMethod;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.not;
+import static net.bytebuddy.matcher.ElementMatchers.returns;
+
+import com.google.auto.service.AutoService;
+import io.opentelemetry.javaagent.tooling.Instrumenter;
+import io.opentelemetry.trace.SpanContext;
+import io.opentelemetry.trace.TracingContextUtils;
+import java.util.Collections;
+import java.util.Map;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.implementation.bytecode.assign.Assigner.Typing;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.apache.logging.log4j.util.SortedArrayStringMap;
+import org.apache.logging.log4j.util.StringMap;
+
+@AutoService(Instrumenter.class)
+public class Log4j27Instrumentation extends Instrumenter.Default {
+  public Log4j27Instrumentation() {
+    super("log4j2", "log4j", "log4j2.7");
+  }
+
+  @Override
+  public ElementMatcher<ClassLoader> classLoaderMatcher() {
+    return not(hasClassesNamed("org.apache.logging.log4j.core.util.ContextDataProvider"));
+  }
+
+  @Override
+  public ElementMatcher<? super TypeDescription> typeMatcher() {
+    return implementsInterface(named("org.apache.logging.log4j.core.ContextDataInjector"));
+  }
+
+  @Override
+  public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
+    return Collections.singletonMap(
+        isMethod()
+            .and(named("injectContextData"))
+            .and(returns(named("org.apache.logging.log4j.util.StringMap"))),
+        Log4j27Instrumentation.class.getName() + "$InjectContextDataAdvice");
+  }
+
+  public static class InjectContextDataAdvice {
+    @Advice.OnMethodExit(suppress = Throwable.class)
+    public static void onExit(
+        @Advice.Return(typing = Typing.DYNAMIC, readOnly = false) StringMap contextData) {
+      SpanContext currentContext = TracingContextUtils.getCurrentSpan().getContext();
+      if (!currentContext.isValid()) {
+        return;
+      }
+
+      if (contextData.containsKey("traceId")) {
+        // Assume already instrumented event if traceId is present.
+        return;
+      }
+
+      StringMap newContextData = new SortedArrayStringMap(contextData);
+      newContextData.putValue("traceId", currentContext.getTraceId().toLowerBase16());
+      newContextData.putValue("spanId", currentContext.getSpanId().toLowerBase16());
+      newContextData.putValue("traceFlags", currentContext.getTraceFlags().toLowerBase16());
```

Comment:
By the way, what do you think about adding some logging constant class in `auto-api`?

Reply:
If they are used in multiple places, why not

Reply:
Done.